src/Data/Decimal.purs

```diff
@@ -0,0 +1,190 @@
+module Data.Decimal
+  ( Decimal()
+  , decimal
+  , getMantissa
+  , getExponent
+  , fromInt
+  , prettyDecimal
+  ) where
+
+import Prelude
+import Data.Array as A
+import Data.Either as E
+import Data.Maybe as M
+import Data.Ord as O
+import Data.String as S
+import Control.Monad.Rec.Class as Rec
+
+import Test.StrongCheck as SC
+import Test.StrongCheck.Gen as Gen
+
+type DecimalR =
+  { mantissa :: Int
+  , exponent :: Int
+  }
+
+-- | An abstract type for exact computing with decimal numbers of finite expansion.
+newtype Decimal = Decimal DecimalR
+
+instance showDecimal :: Show Decimal where
+  show (Decimal d) =
+    "("
+      <> show d.mantissa
+      <> ","
+      <> show d.exponent
+      <> ")"
+
+instance eqDecimal :: Eq Decimal where
+  eq (Decimal d1) (Decimal d2) =
+    d1.mantissa == d2.mantissa
+      && (d1.exponent == d2.exponent || d1.mantissa == 0)
+
+instance ordDecimal :: Ord Decimal where
+  compare (Decimal d1) (Decimal d2) =
+    case compare d1.exponent d2.exponent of
+      EQ -> compare d1.mantissa d2.mantissa
+      o ->
+        if d1.mantissa == 0 && d2.mantissa == 0
+        then EQ
+        else o
+
+type Nat = Int
+
+pow
+  :: Int
+  -> Nat
+  -> Int
+pow i p =
+  Rec.tailRec go { acc: 1, pwr : p }
```

Comment:
@cryogenian If you're confident that the `floor + pow` thing will work properly, then I'm definitely happier to use it than this slow thing :smile:

Just didn't do it because I didn't want to incur the proof obligation myself, since I have no idea.